extensions/amp-fit-text/0.2/fit-text.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import * as styles from './fit-text.css';
+import {omit} from '../../../src/utils/object';
+import {px, setStyle} from '../../../src/style';
+import {useEffect, useRef, useState} from '../../../src/preact';
+
+const {LINE_HEIGHT_EM_} = styles;
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function FitText(props) {
+  const {
+    'children': children,
+    'minFontSize': minFontSize = 6,
+    'maxFontSize': maxFontSize = 72,
+    'width': width = props['style']['width'] || '100%',
+    'height': height = props['style']['height'] || '100%',
+  } = props;
+  const rest = omit(props, [
+    'children',
+    'minFontSize',
+    'maxFontSize',
+    'width',
+    'height',
+  ]);
+  const contentRef = useRef(null);
+  const measurerRef = useRef(null);
+  const [wrapperStyle, setWrapperStyle] = useState(/** @type {Object} */ ({}));
+
+  useEffect(() => {
+    const node = contentRef.current;
+    const observer = new ResizeObserver((entries) => {
+      const last = entries[entries.length - 1];
+      const {height: maxHeight, width: maxWidth} = last['contentRect'];
+      const fontSize = calculateFontSize(
+        measurerRef.current,
+        maxHeight,
+        maxWidth,
+        minFontSize,
+        maxFontSize
+      );
+      const overflownStyle = getOverflowStyle(
+        measurerRef.current,
+        maxHeight,
+        fontSize
+      );
+      setWrapperStyle({...overflownStyle, fontSize: px(fontSize)});
+    });
+    if (node) {
+      observer.observe(node);
+    }
+    return () => observer.unobserve(node);
+  }, [maxFontSize, minFontSize]);
+
+  return (
+    <div {...rest}>
+      <div
+        ref={contentRef}
+        style={{
+          ...styles.fitTextContent,
+          'width': px(width),
```

Comment:
Preact does automatic `px`, right?

Reply:
It appears when I remove `px`, `width`/`height` don't get applied in the element `style` attribute. Alternatively I can change the width and height to `100%` here instead. What do you think?

Reply:
IMHO, for the component itself, this should be entirely up to the client how to achieve the desired size. I.e. we shouldn't even have an opinion on this on React's side. Is this achievable?

Reply:
Overall, is it possible to reduce number of `divs` here? We need to play with this markup I think. I'll look in the storybook. But do you by any chance have a codepen for this?

Reply:
I played around with a number of divs. We can definitely reduce, but what I got out of it, if we add even one layer - we are solidly in the shadow DOM area and it doesn't matter at that point anymore.